### PR TITLE
Add thirdparty library etcpak for faster imports.

### PR DESCRIPTION
**Pull request by fire**
**Add `etcpak` library for faster ETC/ETC2/S3TC imports.**

- `etc` module was renamed to `etcpak` and modified to use the new library.
- PKM importer is removed in the process. Could be re-added if still useful?
- Old library `etc2comp` is removed.
- S3TC compression no longer done via `squish` (but decompression still is).
- Slight modifications to etcpak sources for MinGW compatibility,
  to fix LLVM `-Wc++11-narrowing` errors, and to allow using vendored or
  system libpng.

----

Squish and etc modules have been modified.

Modifications for mingw.

Some times from TPS sample. 

ook ook ook :banana:

Please help measure if this is faster or not.

```
ETCPAK encode took 74 ms
Encoding format: ETC2_RA_AS_RG
ETCPAK encode took 60 ms
Encoding format: DXT5 RGBA8
ETCPAK encode took 74 ms
Encoding format: ETC2_RGB8
ETCPAK encode took 130 ms
Encoding format: DXT1 RGB8
ETCPAK encode took 7 ms
Encoding format: ETC
ETCPAK encode took 30 ms
Encoding format: FORMAT_DXT5_RA_AS_RG
ETCPAK encode took 9 ms
Encoding format: ETC2_RA_AS_RG
ETCPAK encode took 135 ms
Encoding format: DXT5 RGBA8
ETCPAK encode took 8 ms
Encoding format: ETC2_RGB8
```

### CODE REVIEW

**Comment by fire** [comment]
Test using.

```c++
// https://github.com/godotengine/godot/blob/master/editor/editor_file_system.cpp#L1974
for (int i = 0; i < files.size(); i++) {	
  uint64_t t = OS::get_singleton()->get_ticks_msec();
  pr.step(files[i].path.get_file(), i);
  _reimport_file(files[i].path);
  uint16_t final = OS::get_singleton()->get_ticks_msec() - t;
  print_line(vformat("%s import took %f.001 seconds", files[i].path, rtos(final / 1000.0)));
}
```

**Comment by fire** [comment]
From a 8k texture load test, 37 seconds is get_pixel and 13 seconds is texture compression (this pr),

https://github.com/fire/texture-import-godot-project/tree/main/peacock

**Comment by fire** [comment]
Added some flags:

```
env.Prepend(CCFLAGS=["-msse4.1"])
env.Prepend(CCFLAGS=["-mavx"])
env.Prepend(CCFLAGS=["-mavx2"])
env.Prepend(CCFLAGS=["-march=skylake"])
```

File: platform/linuxbsd/detect.py
Comment (Calinou) [inline platform/linuxbsd/detect.py:120]: This will make binaries unable to run on old CPUs (anything made before Skylake). If we want to add support for new instructions, we have to do it in a way that preserves legacy instructions in the binary (which will increase binary size).

File: modules/etcpak/SCsub
Comment (Calinou) [inline modules/etcpak/SCsub:36]: The same CPU compatibility concerns apply here. AVX and AVX2 are unavailable on many low-end Intel CPUs due to market segmentation, even recent models.

Comment (fire) [reply modules/etcpak/SCsub:36]: @reduz said that's ok to use avx as the minimum, and skylake apparently is the oldest win10 supported cpu architecture.

That said let me look at the march list.
Sandybridge is the first gpu that allowed AVX.

> ‘-march=sandybridge’
> Intel Sandy Bridge CPU with 64-bit extensions, MMX, SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2, POPCNT, AVX, AES and PCLMUL instruction set support.

Thoughts on switching to ` env.Prepend(CCFLAGS=["-march=skylake"])` and only `__AVX__` and SSE4.1?

Comment (fire) [reply modules/etcpak/SCsub:36]: You're right it seems like we can only use SSE4.1.

Comment (reduz) [reply modules/etcpak/SCsub:36]: I wonder how much effort would it take to change the code to support multiple versions

Comment (akien-mga) [reply modules/etcpak/SCsub:36]: AFAIK passing `__SSE*__` or `__AVX*__` defines is bogus in any case. Those are defined by the compiler for the target platform, they're not meant to be passed manually. They should indeed be enabled using `-march`, etc., but that means we drop support for any architecture which doesn't support those extensions (i.e. Godot will crash on those platforms).

This of course means killing 32-bit support, but also support for some x86_64 platforms. And that won't work on aarch64 either.

I would suggest making an integration that doesn't attempt to use intrinsics initially, and look at those in a second step. This should be done properly (e.g. by @hpvb who knows a lot about this stuff).

Comment (fire) [reply modules/etcpak/SCsub:36]: That's easy to do, thanks. We can do it as a two stage process.

Comment (fire) [reply modules/etcpak/SCsub:36]: My goals for the next free time slot is to remove march and the defines for instrincs. I'll post the new performance numbers, but I don't want to retest Godot master.

**Comment by fire** [comment]
I don't have the full numbers because there's another pr for better import time logging but here they are.

ETCPAK encode took 920 ms (dxt5)

ETCPAK encode took 3130 ms (etc2)

File: COPYRIGHT.txt
Comment (akien-mga) [inline COPYRIGHT.txt:144]: We don't include emails for other copyright statements in that file so it's probably best to do the same here:
```suggestion
Copyright: 2013-2021, Bartosz Taudul
```

File: modules/etc/SCsub
Comment (akien-mga) [inline modules/etc/SCsub:13]: We should also remove etc2comp since we no longer use it.

File: modules/etcpak/image_etc.h
Comment (akien-mga) [inline modules/etcpak/image_etc.h:31]: No `#pragma once`.

File: thirdparty/README.md
Comment (akien-mga) [inline thirdparty/README.md:109]:
```suggestion
- License: BSD-3-Clause
```

File: thirdparty/README.md
Comment (akien-mga) [inline thirdparty/README.md:111]: Please add the vendoring howto like done for other dependencies:
```
Files extracted from upstream source:

- blabla
```

File: modules/squish/config.py
Comment (akien-mga) [inline modules/squish/config.py:2]: Does this mean etcpak supersedes squish and we can drop it too? If so, we should drop it.

**Review by akien-mga — CHANGES_REQUESTED** [review]
Looks fine overall vendoring-wise, @reduz should review the actual implementation details.

Please vendor the mingw patch (https://github.com/godotengine/godot/commit/4c5193434454126a814acc45453ff48ca7e3a0ee.patch) in `thirdparty/etcpak/patches/`, we'll likely need to drag it for a while until mingw support is added upstream.

**Comment by akien-mga** [comment]
> etc2 import with quicketc2 modified optimization which need to be up streamed.

Does this PR include the QuickETC2 patch? Because that's not documented and it says it's on https://github.com/wolfpld/etcpak/commits/403d38b3f1cb347c196d845d0a05e44a00d17169 which is upstream `master` without QuickETC2.

File: modules/squish/config.py
Comment (fire) [reply modules/squish/config.py:2]: Squish provides DXT1, DXT3, DXT5, BC4 and BC5.

Since all we use is dxt3, dxt5 and etc1, etc2 I disabled it to see if it mattered.

For BC6 we can use Betsy but I still couldn't find a good BC7 compressor.

**Comment by fire** [comment]
We're not using QuickEtc2.

Making other changes.

File: modules/squish/config.py
Comment (fire) [reply modules/squish/config.py]: I'm using etcpak to encode and squish to decode because the changes required more engineering to expose etcpak's decoder.

**Comment by fire** [comment]
I added the copyright from jtbr's bswap and bswap64.

**Comment by fire** [comment]
I'll try to get this error fixed upstream.

```
type 'int' to 'uint32_t' type narrowing.
```

https://github.com/wolfpld/etcpak/issues/11

File: modules/etcpak/image_etc.cpp
Comment (akien-mga) [inline modules/etcpak/image_etc.cpp:48]: I don't think you're actually using `getopt.h` here.

Actually, the `getopt` and `Application.cpp` code should be removed as we don't compile it, so we don't need to ship it.

File: modules/etcpak/image_etc.cpp
Comment (akien-mga) [inline modules/etcpak/image_etc.cpp:35]: Style nitpick, there should be a line after the main header:
```suggestion
#include "image_etc.h"

#include "core/io/image.h"
#include "core/os/copymem.h"
#include "core/os/os.h"
#include "core/string/print_string.h"
```

See https://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html#header-includes

File: thirdparty/etcpak/patches/0001-Make-mingw-deal-with-intrinsics-better-and-narrowing.patch
Comment (akien-mga) [inline thirdparty/etcpak/patches/0001-Make-mingw-deal-with-intrinsics-better-and-narrowing.patch:44]: For the reference, we have our own definitions for bswap in Godot which are probably trivial enough to be considered public domain, and could be used without adding another copyright and license to etcpak:
https://github.com/godotengine/godot/blob/8b6e3d63750db5eab94b5d4e5dc182a678c1e455/core/typedefs.h#L210-L230

I didn't check if they're strictly equivalent.

File: thirdparty/etcpak/patches/0001-Make-mingw-deal-with-intrinsics-better-and-narrowing.patch
Comment (akien-mga) [inline thirdparty/etcpak/patches/0001-Make-mingw-deal-with-intrinsics-better-and-narrowing.patch:193]: This is not blocking for this PR, but I think *BSD systems won't be able to build etcpak as is due to this. That's something we'll have to fix before the 4.0 release as we support those platforms.

See this code, it's pretty tricky to have cross-platform:
https://github.com/godotengine/godot/blob/8b6e3d63750db5eab94b5d4e5dc182a678c1e455/drivers/unix/thread_posix.cpp#L42-L70

**Comment by hpvb** [comment]
I was able to build etcpak on mingw with only the following diff:
```
diff --git a/ProcessRGB.cpp b/ProcessRGB.cpp
index 220d5c5..b9e90a8 100644
--- a/ProcessRGB.cpp
+++ b/ProcessRGB.cpp
@@ -1,5 +1,6 @@
 #include <array>
 #include <string.h>
+#include <limits>
 
 #ifdef __ARM_NEON
 #  include <arm_neon.h>
diff --git a/System.cpp b/System.cpp
index 1383d0e..a09b289 100644
--- a/System.cpp
+++ b/System.cpp
@@ -35,7 +35,7 @@ unsigned int System::CPUCores()
 
 void System::SetThreadName( std::thread& thread, const char* name )
 {
-#ifdef _WIN32
+#ifdef _MSC_VER
     const DWORD MS_VC_EXCEPTION=0x406D1388;
 
 #  pragma pack( push, 8 )
```
This was a cross-build so I will try on native windows as well, but I don't think we need quite as many chances to etcpak to make it work for us.

**Comment by akien-mga** [comment]
@hpvb With your changes (and without https://github.com/godotengine/godot/pull/47370/commits/853b03548de68d66749809df355e56817b74d1d3) I still get an error trying to include `byteswap.h` here:
https://github.com/wolfpld/etcpak/blob/403d38b3f1cb347c196d845d0a05e44a00d17169/ProcessRGB.cpp#L24-L30

Removing this `#else` then works fine as what we want for MinGW/GCC is to use the `__builtin_bswap32` and `__builtin_bswap64` which are aliases right below. Might break other platforms, not so sure what's the exact intent of this code since upstream seems to use `_WIN32` and `_MSC_VER` interchangeably.

**Comment by akien-mga** [comment]
I made a branch based on this PR, with the second commit removed, and replaced by the above patch (minus the addition of `<limits>` which didn't seem needed in the end?) + some other packaging cleanups. WDYT?

**Comment by fire** [comment]
I'm not able to check the code by running it, but from a textual review, the code follows logically.

**Comment by akien-mga** [comment]
I made a version with both commits squashed together and the commit message edited to add some more details (please confirm that I understood and described the changes correctly):

https://github.com/akien-mga/godot/tree/etcpak-cleaned-up

I'll try to do some tests building on Windows with MSVC and MinGW to confirm that it works OK.

**Comment by fire** [comment]
The commit message is correct.

Minor comments:

1. lz4/lz4.c What's the difference between lz4 and minil4?
2. I also removed an etc framing format -- pkm. I've not heard of people using it.

**Comment by hpvb** [comment]
> I made a branch based on this PR, with the second commit removed, and replaced by the above patch (minus the addition of `<limits>` which didn't seem needed in the end?) + some other packaging cleanups. WDYT?
> 
> https://github.com/akien-mga/godot/tree/etcpak-fixup

`<limits>` is really missing and it will break when we update to gcc 10 on mingw.

**Comment by akien-mga** [comment]
Added `<limits>` back. Updated my branch with further fixes to the `SCsub`: https://github.com/akien-mga/godot/tree/etcpak-cleaned-up

I clarified the dependency on libpng, and remove the two defines (`NO_GZIP` and `PNG_INTEL_SSE`) which would likely conflict with our zlib and libpng build depending on what the build order is (redefining defines which impact a header which has possibly already been parsed likely doesn't work reliably). We can re-evaluate them if they're relevant to enable for the whole engine, but not only for etcpak's sake.

Now I see the narrowing errors on the CI which are specific to LLVM, not to mingw:
https://github.com/akien-mga/godot/runs/2325019211?check_suite_focus=true

So I added back @fire's changes with a dedicated patch (separate from the mingw patch). I don't know if they're the proper way to fix the issue, it's just what LLVM recommends (so I assume it's just making explicit what happens otherwise - doesn't mean that it's not an actual bug that should be fixed upstream).

**Comment by hpvb** [comment]
OK, building with native clang on Linux I need to do the following changes, which is closer to ifire's original work:
```diff
diff --git a/BlockData.cpp b/BlockData.cpp
index a2cd032..78f0f55 100644
--- a/BlockData.cpp
+++ b/BlockData.cpp
@@ -340,10 +340,10 @@ static etcpak_force_inline void DecodeT( uint64_t block, uint32_t* dst, uint32_t
     const auto c3b = clampu8( cb1 - table59T58H[codeword] );
 
     const uint32_t col_tab[4] = {
-        cr0 | ( cg0 << 8 ) | ( cb0 << 16 ) | 0xFF000000,
-        c2r | ( c2g << 8 ) | ( c2b << 16 ) | 0xFF000000,
-        cr1 | ( cg1 << 8 ) | ( cb1 << 16 ) | 0xFF000000,
-        c3r | ( c3g << 8 ) | ( c3b << 16 ) | 0xFF000000
+        uint32_t(cr0 | ( cg0 << 8 ) | ( cb0 << 16 ) | 0xFF000000),
+        uint32_t(c2r | ( c2g << 8 ) | ( c2b << 16 ) | 0xFF000000),
+        uint32_t(cr1 | ( cg1 << 8 ) | ( cb1 << 16 ) | 0xFF000000),
+        uint32_t(c3r | ( c3g << 8 ) | ( c3b << 16 ) | 0xFF000000)
     };
 
     const uint32_t indexes = ( block >> 32 ) & 0xFFFFFFFF;
@@ -395,10 +395,10 @@ static etcpak_force_inline void DecodeTAlpha( uint64_t block, uint64_t alpha, ui
     const auto c3b = clampu8( cb1 - table59T58H[codeword] );
 
     const uint32_t col_tab[4] = {
-        cr0 | ( cg0 << 8 ) | ( cb0 << 16 ),
-        c2r | ( c2g << 8 ) | ( c2b << 16 ),
-        cr1 | ( cg1 << 8 ) | ( cb1 << 16 ),
-        c3r | ( c3g << 8 ) | ( c3b << 16 )
+        uint32_t(cr0 | ( cg0 << 8 ) | ( cb0 << 16 )),
+        uint32_t(c2r | ( c2g << 8 ) | ( c2b << 16 )),
+        uint32_t(cr1 | ( cg1 << 8 ) | ( cb1 << 16 )),
+        uint32_t(c3r | ( c3g << 8 ) | ( c3b << 16 ))
     };
 
     const uint32_t indexes = ( block >> 32 ) & 0xFFFFFFFF;
@@ -442,10 +442,10 @@ static etcpak_force_inline void DecodeH( uint64_t block, uint32_t* dst, uint32_t
     const auto codeword = codeword_hi | codeword_lo;
 
     const uint32_t col_tab[] = {
-        clampu8( r0 + table59T58H[codeword] ) | ( clampu8( g0 + table59T58H[codeword] ) << 8 ) | ( clampu8( b0 + table59T58H[codeword] ) << 16 ),
-        clampu8( r0 - table59T58H[codeword] ) | ( clampu8( g0 - table59T58H[codeword] ) << 8 ) | ( clampu8( b0 - table59T58H[codeword] ) << 16 ),
-        clampu8( r1 + table59T58H[codeword] ) | ( clampu8( g1 + table59T58H[codeword] ) << 8 ) | ( clampu8( b1 + table59T58H[codeword] ) << 16 ),
-        clampu8( r1 - table59T58H[codeword] ) | ( clampu8( g1 - table59T58H[codeword] ) << 8 ) | ( clampu8( b1 - table59T58H[codeword] ) << 16 )
+        uint32_t(clampu8( r0 + table59T58H[codeword] ) | ( clampu8( g0 + table59T58H[codeword] ) << 8 ) | ( clampu8( b0 + table59T58H[codeword] ) << 16 )),
+        uint32_t(clampu8( r0 - table59T58H[codeword] ) | ( clampu8( g0 - table59T58H[codeword] ) << 8 ) | ( clampu8( b0 - table59T58H[codeword] ) << 16 )),
+        uint32_t(clampu8( r1 + table59T58H[codeword] ) | ( clampu8( g1 + table59T58H[codeword] ) << 8 ) | ( clampu8( b1 + table59T58H[codeword] ) << 16 )),
+        uint32_t(clampu8( r1 - table59T58H[codeword] ) | ( clampu8( g1 - table59T58H[codeword] ) << 8 ) | ( clampu8( b1 - table59T58H[codeword] ) << 16 ))
     };
 
     for( uint8_t j = 0; j < 4; j++ )
@@ -489,10 +489,10 @@ static etcpak_force_inline void DecodeHAlpha( uint64_t block, uint64_t alpha, ui
     const auto tbl = g_alpha[(alpha >> 48) & 0xF];
 
     const uint32_t col_tab[] = {
-        clampu8( r0 + table59T58H[codeword] ) | ( clampu8( g0 + table59T58H[codeword] ) << 8 ) | ( clampu8( b0 + table59T58H[codeword] ) << 16 ),
-        clampu8( r0 - table59T58H[codeword] ) | ( clampu8( g0 - table59T58H[codeword] ) << 8 ) | ( clampu8( b0 - table59T58H[codeword] ) << 16 ),
-        clampu8( r1 + table59T58H[codeword] ) | ( clampu8( g1 + table59T58H[codeword] ) << 8 ) | ( clampu8( b1 + table59T58H[codeword] ) << 16 ),
-        clampu8( r1 - table59T58H[codeword] ) | ( clampu8( g1 - table59T58H[codeword] ) << 8 ) | ( clampu8( b1 - table59T58H[codeword] ) << 16 )
+        uint32_t(clampu8( r0 + table59T58H[codeword] ) | ( clampu8( g0 + table59T58H[codeword] ) << 8 ) | ( clampu8( b0 + table59T58H[codeword] ) << 16 )),
+        uint32_t(clampu8( r0 - table59T58H[codeword] ) | ( clampu8( g0 - table59T58H[codeword] ) << 8 ) | ( clampu8( b0 - table59T58H[codeword] ) << 16 )),
+        uint32_t(clampu8( r1 + table59T58H[codeword] ) | ( clampu8( g1 + table59T58H[codeword] ) << 8 ) | ( clampu8( b1 + table59T58H[codeword] ) << 16 )),
+        uint32_t(clampu8( r1 - table59T58H[codeword] ) | ( clampu8( g1 - table59T58H[codeword] ) << 8 ) | ( clampu8( b1 - table59T58H[codeword] ) << 16 ))
     };
 
     for( uint8_t j = 0; j < 4; j++ )
diff --git a/ProcessRGB.cpp b/ProcessRGB.cpp
index 220d5c5..b9e90a8 100644
--- a/ProcessRGB.cpp
+++ b/ProcessRGB.cpp
@@ -1,5 +1,6 @@
 #include <array>
 #include <string.h>
+#include <limits>
 
 #ifdef __ARM_NEON
 #  include <arm_neon.h>
diff --git a/System.cpp b/System.cpp
index 1383d0e..a09b289 100644
--- a/System.cpp
+++ b/System.cpp
@@ -35,7 +35,7 @@ unsigned int System::CPUCores()
 
 void System::SetThreadName( std::thread& thread, const char* name )
 {
-#ifdef _WIN32
+#ifdef _MSC_VER
     const DWORD MS_VC_EXCEPTION=0x406D1388;
 
 #  pragma pack( push, 8 )
 ```

**Review by akien-mga — APPROVED** [review]
I updated this PR with some improvements as discussed above, thanks @fire and @hpvb for the help.

I think it's in a good state now if it passes CI.

I'll see if I can get some import performance metrics before merging, so that we can see the difference between etc2comp.

**Comment by fire** [comment]
https://github.com/fire/godot/tree/importer-timings can be helpful

**Comment by akien-mga** [comment]
Quick test of importing all assets of the TPS demo (note: includes some DAEs and glTFs too):

- Before this PR: 6m30s
- After this PR: 1m13s

Example:

- Before:
```
res://enemies/red_robot/textures/red_robot_albedo.png import took 2.871 seconds
res://enemies/red_robot/textures/red_robot_emission.png import took 1.115 seconds
res://enemies/red_robot/textures/red_robot_normal.png import took 3.756 seconds
res://enemies/red_robot/textures/red_robot_orm.png import took 3.275 seconds
```

- After:
```
res://enemies/red_robot/textures/red_robot_normal.png import took 0.869 seconds
res://enemies/red_robot/textures/red_robot_emission.png import took 0.489 seconds
res://enemies/red_robot/textures/red_robot_albedo.png import took 0.659 seconds
res://enemies/red_robot/textures/red_robot_orm.png import took 0.797 seconds
```

So there's already a pretty significant gain on this very unscientific test. This can likely be improved further by ensuring that we properly use the available SSE4.1 / AVX2 code paths on compatible CPUs.

**Comment by akien-mga** [comment]
Thanks!

**Comment by fire** [comment]
```
#elif !defined(__APPLE__) && !defined(__MINGW32__ )
    pthread_setname_np( thread.native_handle(), name );
#endif
```

LLVM + MINGW still requires the __MINGW32__  check.

**Comment by akien-mga** [comment]
> ```
> #elif !defined(__APPLE__) && !defined(__MINGW32__ )
>     pthread_setname_np( thread.native_handle(), name );
> #endif
> ```
> 
> LLVM + MINGW still requires the `__MINGW32__` check.

That doesn't seem like a correct fix. What's the error exactly?

This API is not compiler dependent, it's likely related to whether your MinGW distro is configured to use pthread or another thread API. Typically MinGW-GCC distros come with winpthreads, there might be a switch or a package you need for mingw-llvm.

File: thirdparty/etcpak/patches/0001-Make-mingw-deal-with-intrinsics-better-and-narrowing.patch
Comment (akien-mga) [reply thirdparty/etcpak/patches/0001-Make-mingw-deal-with-intrinsics-better-and-narrowing.patch]: Should be fixed with #47852 (doing like upstream did for Apple - no thread names on *BSD).

**Comment by akien-mga** [comment]
If anyone is interested, I did some work to backport this change (and follow-ups) to the `3.x` branch, which could help speed import times significantly:

https://github.com/akien-mga/godot/commits/3.x-etcpak

It's WIP and I don't plan to work further on it as it turns out etcpak doesn't support all formats needed for GLES2 support in the `3.x` branch, so a hybrid solution would be needed (keeping the old importers for the formats that etcpak doesn't support).

**Pull request merged**
